File: Day 12/Python/sol.py
"""
FILE: Day 12/Python/sol.py
Author: Alex Jones
Desc: Solution to day 12 problems (23 & 24) for Advent of Code 2025, solved in Python 3.
"""
data = open("Day 12/data.txt", "r").read().rstrip().replace("\r","").split("\n\n")

# Naive solution that only excludes solutions that can't possibly fit,
# but this also happens to be all the real inputs for this puzzle.
sizes = [len([c for c in s if c == '#']) for s in data[:-1]]
total = 0
for region in data[-1].split("\n"):
    (area, num_shapes) = region.split(": ")
    (area, num_shapes) = (area.split("x"), num_shapes.split())
    area = int(area[0]) * int(area[1])
    min_area = sum(int(n) * s for n, s in zip(num_shapes, sizes))
    total += min_area <= area
print("Problem 23:", total)
print("Problem 24: Christmas (Spirit) Freebie :)")

File: Day 12/Rust/sol.rs
/*
 * FILE: Day 12/Rust/sol.rs
 * Author: Alex Jones
 * Desc: Solution to day 12 problems (23 & 24) for Advent of Code 2025, solved in Rust.
 */
use std::fs;

use itertools::Itertools;

pub fn solve() {
    let contents: String = fs::read_to_string("Day 12/data.txt")
        .expect("Could not read data file")
        .trim()
        .replace("\r", "");
    let data: Vec<&str> = contents.split("\n\n").collect();

    let shape_sizes = data
        .iter()
        .take(data.len() - 1)
        .map(|s| s.chars().filter(|&c| c == '#').count())
        .collect::<Vec<_>>();
    println!(
        "Problem 23: {}",
        data.last()
            .unwrap()
            .split("\n")
            .filter(|region| {
                let parts: (&str, &str) = region.split(": ").collect_tuple().unwrap();
                let area: u32 = parts
                    .0
                    .split("x")
                    .map(|n| n.parse::<u32>().unwrap())
                    .product();
                let min_area: u32 = parts
                    .1
                    .split_whitespace()
                    .map(|n| n.parse::<u32>().unwrap())
                    .zip(shape_sizes.iter())
                    .map(|(num, size)| num * (*size as u32))
                    .sum();
                min_area <= area
            })
            .count()
    );
    println!("Problem 24: Christmas (Spirit) Freebie :)");
}
